src/librustc_metadata/tyencode.rs

```diff
@@ -407,7 +407,14 @@ pub fn enc_existential_bounds<'a,'tcx>(w: &mut Cursor<Vec<u8>>,
 
     enc_region(w, cx, bs.region_bound);
 
-    for tp in &bs.projection_bounds {
+    // Encode projection_bounds in a stable order
+    let mut projection_bounds: Vec<_> = bs.projection_bounds
+                                          .iter()
+                                          .map(|b| (b.item_name().as_str(), b))
+                                          .collect();
+    projection_bounds.sort_by_key(|&(ref name, _)| name.clone());
```

Comment:
Newbie question (and total nit): why couldn't this basically be the following (or something close to it)?
```rust
let mut projection_bounds = bs.projection_bounds
                              .clone()
                              .sort_by_key(|&b| b.item_name().as_str());
```

Because then it seems we'd be able to keep the next line just `for tp in &projection_bounds { ... }`, avoiding the extra call to `map`.

And even if we didn't do this the next line still seems like it could stay `for tp in &projection_bounds` with `s/tp.0/tp.1.0` in the loop body.